master/internal/api_user.go

```diff
@@ -683,3 +696,253 @@ func (a *apiServer) PostUserActivity(
+		if err != nil || d < 0 {
+			return nil, status.Error(codes.InvalidArgument,
+				"Lifespan must be a Go-formatted duration string with a positive value")
+		}
+		tokenExpiration = d
+	}
+
+	token, err := user.RevokeAndCreateLongLivedToken(
+		ctx, targetFullUser.ID, user.WithTokenExpiry(&tokenExpiration))
+	if err != nil {
+		return nil, err
+	}
+	return &apiv1.PostUserLongLivedTokenResponse{LongLivedToken: token}, nil
+}
+
+// GET active Long lived token returns token information for the logged in user.
+func (a *apiServer) GetLongLivedToken(
+	ctx context.Context, req *apiv1.GetLongLivedTokenRequest,
+) (*apiv1.GetLongLivedTokenResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if err = user.AuthZProvider.Get().CanGetUsersOwnToken(ctx, *curUser); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	tokenInfo, err := user.GetLongLivedTokenInfo(ctx, curUser.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &apiv1.GetLongLivedTokenResponse{LongLivedTokenInfo: tokenInfo.Proto()}, nil
+}
+
+// GET All (active / revoked) Long lived token returns all long-lived token information from user_session db.
+func (a *apiServer) GetAllLongLivedTokens(
+	ctx context.Context, req *apiv1.GetAllLongLivedTokensRequest,
+) (*apiv1.GetAllLongLivedTokensResponse, error) {
+	sortColMap := map[apiv1.GetAllLongLivedTokensRequest_SortBy]string{
+		apiv1.GetAllLongLivedTokensRequest_SORT_BY_UNSPECIFIED:       "id",
+		apiv1.GetAllLongLivedTokensRequest_SORT_BY_USER_ID:           "user_id",
+		apiv1.GetAllLongLivedTokensRequest_SORT_BY_EXPIRY:            "expiry",
+		apiv1.GetAllLongLivedTokensRequest_SORT_BY_CREATED_AT:        "created_at",
+		apiv1.GetAllLongLivedTokensRequest_SORT_BY_TOKEN_TYPE:        "token_type",
+		apiv1.GetAllLongLivedTokensRequest_SORT_BY_IS_REVOKED:        "is_revoked",
+		apiv1.GetAllLongLivedTokensRequest_SORT_BY_TOKEN_DESCRIPTION: "token_description",
+	}
+	orderByMap := map[apiv1.OrderBy]string{
+		apiv1.OrderBy_ORDER_BY_UNSPECIFIED: "ASC",
+		apiv1.OrderBy_ORDER_BY_ASC:         "ASC",
+		apiv1.OrderBy_ORDER_BY_DESC:        "DESC",
+	}
+
+	userSessions := []model.UserSession{}
+
+	query := db.Bun().NewSelect().Model(&userSessions).
+		ModelTableExpr("user_sessions as us").
+		Column("us.id").
+		Column("us.user_id").
+		Column("us.expiry").
+		Column("us.created_at").
+		Column("us.token_type").
+		Column("us.is_revoked").
+		Column("us.token_description")
+
+	if req.Name != "" {
+		var userIDForGivenUsername model.UserID
+		nameFilterExpr := "%" + req.Name + "%"
+		err := db.Bun().NewSelect().
+			Table("users").
+			Column("id").
+			Where("username ILIKE ?", nameFilterExpr).
+			Scan(ctx, &userIDForGivenUsername)
+		if err != nil {
+			return nil, err
+		}
+		query.Where("us.user_id = ?", userIDForGivenUsername)
+	}
+	if req.TokenType != nil && req.TokenType.String() != userv1.TokenType_TOKEN_TYPE_UNSPECIFIED.String() {
```

Comment:
I'm not sure we need to allow for filtering on the TokenType for this endpoint, since we're expecting that the use case here is for Long Lived Access Tokens specifically.

Reply:
That being said, it could be interesting to eventually support this in a One Token approach!

Reply:
You make a good point that we need to support it in future approach. So, instead of removing this all together, do you think having this option exposed to the customer with a description of it as "currently only supporting Long Lived Access Tokens" specifically, and maybe default the drop down to LONG_LIVED_TOKEN (if possible) makes sense?

Just a thought

Reply:
Actually have removed it till we need this filter.